### ISSUE TITLE
DefaultMavenPublication#determineMainArtifact behaves very badly with "pom" packaging

### ISSUE THREAD

**Issue by Vampire** [gradle/gradle#4985]
The aforementioned method searches all unclassified artifacts with extension.
If it finds one, it is considered the main artifact unconditionally.
If if finds multiple and the extension of one matches the packaging, this is considered the main artifact.
If it finds none or multiple and none matches, none is considered the main artifact.

Now the problem is, if you have `pom` packaging and one other unclassified artifact with extension.
In my case this comes from
```groovy
artifact(file("$buildDir/publications/$publication.name/pom-default.xml.asc")) {
    extension 'pom.asc'
    builtBy signArchives
}
```

In that case this artifact is considered the main artifact and its file is taken for the `pom` main artifact, then at publishing time first the meant `pom.asc` file is uploaded to the repository, then the separately tracked `pom` file overwrites it.

I didn't find a work-around yet that does not involve uploading a dummy file without classification and with extension to the repository, as then there are two unclassified artifacts and none matches the packaging, or alternatively changing the packaging to `pom.asc` which are both pretty unclean solutions.

I also tried to add the pom additionally as artifact manually, but then the validator complains that there are multiple artifacts with extension `pom` and without classifier.

I also tried to configure in gradle `pom.asc` as packaging and then before the signing task replace the packaging in the generated POM file manually, unfortunately at that time the packaging is not taken from the Gradle configuration but parsed from the actual POM file.

If you have a better workaround than releasing a dummy file and having a wrong packaging value (besides using the old `maven` plugin), I'd be happy to hear about it.

---

I guess the proper solution in Gradle should be to return `null` from `DefaultMavenPublication#determineMainArtifact`, if the packaging is `pom`, as then the separately held POM is the main artifact itself.

**Comment by Vampire** [comment]
This is even worse, the nice new `IMPROVED_POM_SUPPORT` does not work properly if the packaging of the downloaded POM is not one of `jar`, `ejb`, `bundle`, `maven-plugin`, `eclipse-plugin`, so using the "wrong packaging work-around" destroys the compile / runtime separation from POM. :-(

**Comment by marcphilipp** [comment]
Thanks for reporting this issue, @Vampire! I am currently working on adding explicit support for publications to the signing plugin in #4943. This way, you won't have to jump through hoops to publish signature files in addition to the main artifacts. As this will very likely solve your problem and change the internal structure of the code, I'm closing this issue for now. I'd appreciate your feedback on #4943 once it's done.

**Comment by Vampire** [comment]
Ah, nice, planned for 4.8.

Actually it was not that hard to do as you read in the wild. 
Basically just
- adding the `signArchives.signatures` to the publication, setting their extension to `signature.type`,
- adding the generated POMs to the `signArchives` task in a `tasks.withType(Sign)` closure,
- adding the POM signature files with hard-coded paths to the publication (in the `withType` the publication was already requested and cannot be modified)

The only real problem I hit was this one.

 ---

Well, regarding the actual topic, your changes will solve my use-case, but if you don't touch the mentioned code, the problem will persist.

It happens with any single non-classifier artifact that is added to a pom-packaging publication. 

If you change the relevant code, it could of course maybe also solve this one.

Could you maybe anyways consider moving this from `feature` to the `bug` it is and incorporating the fix in 4.7 as a preliminary fix?

**Comment by Vampire** [comment]
Unfortunately the `pom.asc` packaging workaround does not work as I just found out.
It does work with maven as you there have to specify the type at consumption time as `pom` or it will always search for the `jar`.
But Gradle always searches for the `jar`, except packaging is set to `pom` as it seems. :-(
